### PR TITLE
Amazon Authenticator

### PR DESCRIPTION
**Pull request by hmrajas**

### CODE REVIEW

File: amazon/org.wso2.carbon.identity.authenticator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    <name>WSO2 Carbon - Authenticator Library For Amazon</name>
+    <url>http://wso2.org</url>
+    <parent>
+        <groupId>org.wso2.carbon</groupId>
+        <artifactId>is-connectors</artifactId>
+        <version>1.0.0</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <dependencies>
+    <dependency>
```
Comment (Shakila) [inline amazon/org.wso2.carbon.identity.authenticator/pom.xml]: You can remove this dependency. It's already in parent pom.

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticatorConstants.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    //Amazon token  endpoint URL
+    public static final String AMAZON_TOKEN_ENDPOINT = "https://api.amazon.com/auth/o2/token";
+    //Amazon user info endpoint URL
+    public static final String AMAZON_USERINFO_ENDPOINT = "https://api.amazon.com/user/profile";
+
+    public static final String CLIENT_ID="Client Id";
+    public static final String CLIENT_SECRET="Client Secret";
+    public static final String CALLBACK_URL="Callback URL";
+    public static final String USER_ID="user_id";
+    public static final String AMAZON_BASIC_SCOPE = "profile";
```
Comment (Shakila) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticatorConstants.java]: Change the name meaningfully.

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            AuthenticatedUser authenticatedUserObj;
+            String json = sendRequest(AmazonAuthenticatorConstants.AMAZON_USERINFO_ENDPOINT,
+                    oAuthResponse.getParam(OIDCAuthenticatorConstants.ACCESS_TOKEN));
+            JSONObject obj = new JSONObject(json);
+            authenticatedUserObj = AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier((String) obj.get(AmazonAuthenticatorConstants.USER_ID));
+            authenticatedUserObj.setAuthenticatedSubjectIdentifier((String) obj.get(AmazonAuthenticatorConstants.USER_ID));
+            claims = getSubjectAttributes(oAuthResponse, authenticatorProperties);
+            authenticatedUserObj.setUserAttributes(claims);
+            context.setSubject(authenticatedUserObj);
+        } catch (OAuthProblemException | IOException | JSONException e) {
```
Comment (Shakila) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: We need to have Java 1.7 or above to use this feature. Please add README and specify the pre-requisites.

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.identity.authenticator.amazon;
+
+
```
Comment (Shakila) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: One empty line enough.

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java (hunk 3 above)
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Added in README

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java (hunk 4 above)
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Space removed

File: amazon/org.wso2.carbon.identity.authenticator/pom.xml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.wso2.carbon.identity.authenticator.amazon</groupId>
+    <artifactId>org.wso2.carbon.identity.authenticator.Amazon</artifactId>
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/pom.xml]: artifact id to simple case https://github.com/wso2/is-connectors/pull/53

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    public String getName() {
+        return AmazonAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    /**
+     * Get Configuration Properties
+     */
+    @Override
+    public List<Property> getConfigurationProperties() {
+
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: space removed in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+     *
+     * @param scope                   Scope
+     * @param authenticatorProperties Authentication properties.
+     * @return OAuth2 Scope
+     */
+    @Override
+    protected String getScope(String scope, Map<String, String> authenticatorProperties) {
+        return AmazonAuthenticatorConstants.AMAZON_SCOPE_PROFILE;
+    }
+
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Description added in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        Property callbackUrl = new Property();
+        callbackUrl.setDisplayName(AmazonAuthenticatorConstants.CALLBACK_URL);
+        callbackUrl.setName(IdentityApplicationConstants.OAuth2.CALLBACK_URL);
+        callbackUrl.setDescription("Enter the callback url");
+        callbackUrl.setDisplayOrder(2);
+        configProperties.add(callbackUrl);
+        return configProperties;
+    }
+
+
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: space removed in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                throw new AuthenticationFailedException("Access token is empty or null");
+            }
+            context.setProperty(OIDCAuthenticatorConstants.ACCESS_TOKEN, accessToken);
+            Map<ClaimMapping, String> claims;
+            AuthenticatedUser authenticatedUserObj;
+            String json = sendRequest(AmazonAuthenticatorConstants.AMAZON_USERINFO_ENDPOINT,
+                    oAuthResponse.getParam(OIDCAuthenticatorConstants.ACCESS_TOKEN));
+            JSONObject obj = new JSONObject(json);
+            authenticatedUserObj = AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier((String) obj.get(AmazonAuthenticatorConstants.USER_ID));
+            authenticatedUserObj.setAuthenticatedSubjectIdentifier((String) obj.get(AmazonAuthenticatorConstants.USER_ID));
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: code format applied in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+            claims = getSubjectAttributes(oAuthResponse, authenticatorProperties);
+            authenticatedUserObj.setUserAttributes(claims);
+            context.setSubject(authenticatedUserObj);
+        } catch (OAuthProblemException | IOException | JSONException e) {
+            throw new AuthenticationFailedException("Authentication process failed", e);
+        }
+    }
+
```
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Method description added in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    /**
+     * Check ID token in Amazon OAuth.
+     */
+    @Override
+    protected boolean requiredIDToken(Map<String, String> authenticatorProperties) {
+        return false;
+    }
+
+    /**
+     * Get the friendly name of the Authenticator
```
Comment (keerthu) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Use "." at the end of the comments"

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        try {
+            oAuthResponse = oAuthClient.accessToken(accessRequest);
+        } catch (OAuthSystemException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Exception while requesting access token", e);
+            }
+            throw new AuthenticationFailedException(e.getMessage(), e);
+        } catch (OAuthProblemException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Exception while requesting access token", e);
```
Comment (keerthu) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Don't we need to throw the exception here?

Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Fixed in https://github.com/wso2/is-connectors/pull/63

File: amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java (hunk 11 above)
Comment (hmrajas) [inline amazon/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/amazon/AmazonAuthenticator.java]: Fixed in https://github.com/wso2/is-connectors/pull/63

**Pull request merged**